src/Integration/LocalServices/ProjectSystemFilter.cs

```diff
@@ -85,13 +72,58 @@ public bool IsAccepted(DteProject dteProject)
 
         public void SetTestRegex(Regex regex)
         {
+            if (regex == null)
+            {
+                throw new ArgumentNullException(nameof(regex));
+            }
+
             this.testRegex = regex;
-            Debug.Assert(this.testRegex == null || (this.testRegex.MatchTimeout != Regex.InfiniteMatchTimeout), "Should have set non-infinite timeout");
+            Debug.Assert(this.testRegex.MatchTimeout != Regex.InfiniteMatchTimeout, "Should have set non-infinite timeout");
         }
 
         #endregion
 
         #region Helpers
+        private static bool IsTestProject(IVsHierarchy projectHierarchy, Regex testProjectNameRegex, string projectName)
+        {
+            // Otherwise, try to detect test project using known project types and/or regex match
+            if (ProjectSystemHelper.IsKnownTestProject(projectHierarchy))
+            {
+                return true;
+            }
+
+            // Heuristically exclude by project name
+            return (testProjectNameRegex != null && testProjectNameRegex.IsMatch(projectName));
+        }
+
+        private static bool IsNotSupportedProject(DteProject project)
+        {
+            var language = Language.ForProject(project);
+            return (language == null || !language.IsSupported);
+        }
+
+        private static bool IsExcludedViaProjectProperty(IVsBuildPropertyStorage propertyStorage)
+        {
+            Debug.Assert(propertyStorage != null);
+
+            // General exclusions
+            // If exclusion property is set, this takes precedence
```

Comment:
Duncan was referring to the first comment i.e. the general exclusion, unless I'm missing something. I've updated the comment